cpp/src/gandiva/llvm_external_ir_store.h

```diff
@@ -0,0 +1,40 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <arrow/status.h>
+#include <gandiva/visibility.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <memory>
+#include <vector>
+
+namespace gandiva {
+using arrow::Status;
+
+class GANDIVA_EXPORT LLVMExternalIRStore {
```

Comment:
Nit, but is "IR" the customary term here? LLVM has textual IR and binary bitcode, is "IR" usable for both?

Reply:
Internally, IR/bitcode terms are used interchangeably in the code base. The current API in `engine.cc` for loading IR is called `LoadPreCompiledIR`, which actually loads `bitcode` instead of textual IR:
```
auto bitcode = llvm::StringRef(reinterpret_cast<const char*>(kPrecompiledBitcode),
                                 kPrecompiledBitcodeSize);
```

But the newly added `LLVMExternalIRStore` is indeed only expected to accept binary encoded bitcode instead of textual IR, so now I rename this class to be `LLVMExternalBitcodeStore` to avoid any confusion.